### ISSUE TITLE
Unable to find consistent type in mnist spec

### ISSUE THREAD

**Issue by ndslusarz** [vehicle-lang/vehicle#390]
Trying to compile modified mnist spec:

```
@network
mnist : Tensor Rat [28, 28] -> Vector Rat 10

--------------------------------------------------------------------------------
-- Definition of robustness around a point

-- First we define the parameter `epsilon` that will represent the radius of the
-- ball that we want the network to be robust in. Note that we declare this as
-- a parameter which allows the value of `epsilon` to be specified at compile
-- time rather than be fixed in the specification.

--epsilon : Rat
--epsilon = 0.01

--delta : Rat
--delta = 0.02

-- Next we define what it means for an image `x` to be in a ball of
-- size epsilon around 0.
boundedByEpsilon : Tensor Rat [28, 28] -> Bool
boundedByEpsilon x = forall i j . -0.01 <= x ! i ! j <= 0.01

--The disjunction avoids having to define absolute value
boundedByDelta : Tensor Rat [28, 28] -> Tensor Rat [28, 28] -> Bool
boundedByDelta x y = forall i . (-0.02 <= (((mnist x) ! i ) - ((mnist y) ! i )) <= 0.02) 
                                or (-0.02 <= (-((mnist x) ! i ) - ((mnist y) ! i )) <= 0.02)
                                or (-0.02 <= (((mnist x) ! i) + ((mnist y) ! i )) <= 0.02)
                                or (-0.02 <= (-((mnist x) ! i ) - ((mnist y) ! i )) <= 0.02)


robustSingleInput : Tensor Rat [28, 28] -> Bool
robustSingleInput image = forall pertubation .
    let perturbedImage = image - pertubation in
    boundedByEpsilon pertubation => True

@property
robust1 : Bool
robust1 = forall x . robustSingleInput x
```

 spec throws an error where it's unable to find a type for `perturbation`

```
"Error during specification compilation: " + result.stderr.decode("UTF-8")  
Exception: Error during specification compilation: Error at Line 44, Columns 32-33: unable to find a consistent type for the overloaded expression '-'. In particular 28 != image.
Fix: check your types
```

**Comment by MatthewDaggitt** [comment]
I'm afraid I don't get this error on `dev` branch at all. Do you get on `dev`? If not, then I'm afraid that means its a change on your personal branch that is responsible.

**Comment by ndslusarz** [comment]
I've double checked now - you're right, after getting a fresh branch from the lastest pull it does type check when calling it straight (calling vehicle compile with TypeCheck only). However the error does appear still if one calls the python script (called `test_mnist.py` on `loss_experiments` branch, I'll copy the code below as well) for this specification which is calling to compile to loss function with all the neccesary arguments. It seems like error originates somewhere in python code then.

```
import random
import numpy as np
import tensorflow as tf
from sklearn.model_selection import train_test_split
import tensorflow.keras as keras
import h5py
from keras.datasets import mnist

from vehicle import generate_loss_function
#from not_init import generate_loss_function

#trainingDataFiles = "vehicle-python/tests/experiments_loss/TrainingData/HCAS_rect_TrainingData_v%d_pra%d_tau%02d.h5"


def train(
    model,
    train_dataset,
    test_dataset,
    epochs,
    alfa,
    beta,
    path_to_spec,
    function_name,
    networks,
    datasets,
    parameters,
    quantifier_sampling,
):
    optimizer = keras.optimizers.Adam()
    ce_batch_loss = keras.losses.BinaryCrossentropy()
    vehicle_batch_loss = generate_loss_function(
        specification=path_to_spec,
        function_name=function_name,
        networks=networks,
        datasets=datasets,
        parameters=parameters,
        quantifier_sampling=quantifier_sampling,
    )

    train_acc_metric = keras.metrics.BinaryCrossentropy()
    test_acc_metric = keras.metrics.BinaryCrossentropy()
    train_loss_metric = keras.metrics.BinaryCrossentropy()
    test_loss_metric = keras.metrics.BinaryCrossentropy()

    for epoch in range(epochs):
        print(f"\nEpoch {epoch + 1}")

        # Iterate over the batches of the dataset.
        for x_batch_train, y_batch_train in train_dataset:
            # Open a GradientTape to record the operations run during the forward pass, which enables auto-differentiation.
            with tf.GradientTape() as tape:
                outputs = model(
                    x_batch_train, training=True
                )  # Outputs for this minibatch
                ce_loss_value = ce_batch_loss(y_batch_train, outputs)
                vehicle_loss = vehicle_batch_loss()
                total_loss = ce_loss_value * alfa + vehicle_loss * beta
            # Use the gradient tape to automatically retrieve the gradients of the trainable variables with respect to the loss.
            grads = tape.gradient(total_loss, model.trainable_weights)
            # Run one step of gradient descent by updating the value of the variables to minimize the loss.
            optimizer.apply_gradients(zip(grads, model.trainable_weights))

        # Run a training loop at the end of each epoch.
        for x_batch_train, y_batch_train in train_dataset:
            train_outputs = model(x_batch_train, training=False)
            train_acc_metric.update_state(y_batch_train, train_outputs)
            train_loss_metric.update_state(y_batch_train, train_outputs)

        # Run a testing loop at the end of each epoch.
        for x_batch_test, y_batch_test in test_dataset:
            test_outputs = model(x_batch_test, training=False)
            test_acc_metric.update_state(y_batch_test, test_outputs)
            test_loss_metric.update_state(y_batch_test, test_outputs)

        train_acc = train_acc_metric.result()
        test_acc = test_acc_metric.result()
        train_loss = train_loss_metric.result()
        test_loss = test_loss_metric.result()

        train_acc_metric.reset_states()
        test_acc_metric.reset_states()
        train_loss_metric.reset_states()
        test_loss_metric.reset_states()

        print(
            f"Train acc: {float(train_acc):.4f}, Train loss: {float(train_loss):.4f} --- Test acc: {float(test_acc):.4f}, Test loss: {float(test_loss):.4f}"
        )

    return model


if __name__ == "__main__":
    path_to_spec = "vehicle-python/tests/mnist.vcl"
    function_name = "robust1"
    model = keras.Sequential(
        [
            keras.layers.Input(shape=(5,)),
            keras.layers.Dense(10),
            keras.layers.Dense(5),
        ]
    )
    networks = {"mnist": model}

    quantifier_sampling = {"x": lambda: random.uniform(0.5, 0.5)}

    batch_size = 1
    epochs = 4
    alfa = 0
    beta = 1

    (X_train, y_train), (X_test, y_test) = mnist.load_data()


    train_dataset = tf.data.Dataset.from_tensor_slices((X_train, y_train))
    test_dataset = tf.data.Dataset.from_tensor_slices((X_test, y_test))

    train_dataset = train_dataset.shuffle(buffer_size=1024).batch(batch_size)
    test_dataset = test_dataset.batch(batch_size)

    model = train(
        model,
        train_dataset,
        test_dataset,
        epochs,
        alfa,
        beta,
        path_to_spec,
        function_name,
        networks,
        {},
        {},
        quantifier_sampling,
    )

```

**Comment by wenkokke** [comment]
Is it possible that the Python code is using an older Vehicle version that is installed on your computer?
